aten/src/ATen/core/type.cpp

```diff
@@ -523,6 +523,25 @@ TensorTypePtr TensorType::merge(TensorTypePtr other, bool merge_sizes) const {
       undef);
 }
 
+template <typename T>
+bool is_null_or_equal(c10::optional<T> a, c10::IntArrayRef b) {
+  return !a.has_value() || a.value() == b;
+}
+
+bool TensorType::matchTensor(const at::Tensor& t) {
+  bool undef = undefined().value_or(true);
+  if (undef && !t.defined()) {
+      return true;
+  }
+  bool rg = at::GradMode::is_enabled() && t.requires_grad();
+  return t.defined() && !undef && isComplete()
```

Comment:
I feel like with `isComplete` we will be failing a lot of guards incorrectly as we start using dynamic shapes more and more. Since we the dynamic shapes a tensor type won't be `Complete` but it might still match an incoming tensor type.
Could you please at least insert an assert here `TORCH_INTERNAL_ASSERT(isComplete())` ? 
Otherwise we will start silently failing guards as soon as we get any dynamic shapes.